File: Answer/1390.四因数.cs
/*
 * @lc app=leetcode.cn id=1390 lang=csharp
 *
 * [1390] 四因数
 */

// @lc code=start
public class Solution
{
    public int SumFourDivisors(int[] nums)
    {
        var res = 0;
        foreach (var n in nums)
        {
            var di = AllDivisors(n);
            if (di.Length == 4)
            {
                res += di.Sum();
            }
        }
        return res;

        int[] AllDivisors(int n)
        {
            var res = new HashSet<int>();
            for (int i = 1; i <= Math.Sqrt(n); i++)
            {
                if (n % i == 0)
                {
                    res.Add(i);
                    res.Add(n / i);
                }
            }
            return res.ToArray();
        }
    }
}
// @lc code=end

File: Answer/1390.四因数.java
/*
 * @lc app=leetcode.cn id=1390 lang=java
 *
 * [1390] 四因数
 */

// @lc code=start
import java.util.HashSet;

class Solution {

    public int sumFourDivisors(int[] nums) {
        var res = 0;
        for (var n : nums) {
            var di = AllDivisors(n);
            if (di.size() == 4) {
                for (var i : di) {
                    res += i;
                }
            }
        }
        return res;
    }

    HashSet<Integer> AllDivisors(int n) {
        HashSet<Integer> res = new HashSet<>();
        for (int i = 1; i <= Math.sqrt(n); i++) {
            if (n % i == 0) {
                res.add(i);
                res.add(n / i);
            }
        }
        return res;
    }
}
// @lc code=end

